python/cuml/tsa/arima.pyx

```diff
@@ -0,0 +1,817 @@
+        # fit. The "best" model is probably a p=1, so we will assume that for now.
+        p_best = 1
+        x = np.zeros((len(y) - p_best, p_best))
+        # create lagged series set
+        for lag in range(1, p_best+1):
+            # create lag and trim appropriately from front so they are all the same size
+            x[:, lag-1] = y[p_best-lag:-lag].T
+
+        # LS fit a*X - Y
+        y_ar = y[p_best:]
+
+        (ar_fit, _, _, _) = np.linalg.lstsq(x, y_ar.T, rcond=None)
+        # print("initial_ar_fit:", ar_fit)
+        # set_trace()
+        if q == 0:
+            params_init[d:] = ar_fit
+        else:
+            residual = y[p_best:] - np.dot(x, ar_fit)
+
+            assert p >= p_best
+            p_diff = p - p_best
+
+            x_resid = np.zeros((len(residual) - q - p_diff, q))
+            x_ar2 = np.zeros((len(residual) - q - p_diff, p))
+
+            # create lagged residual and ar term
+            for lag in range(1, q+1):
+                x_resid[:, lag-1] = (residual[q-lag:-lag].T)[p_diff:]
+            for lag in range(1, p+1):
+                x_ar2[:, lag-1] = (y[p-lag:-lag].T)[q:]
+
+            X = np.column_stack((x_ar2, x_resid))
+            (arma_fit, _, _, _) = np.linalg.lstsq(X, y_ar[(q+p_diff):].T, rcond=None)
+
+            params_init[d:] = arma_fit
+
+    else:
+        # case when p == 0 and q>0
+
+        # when p==0, MA params are often -1
+        # TODO: See how `statsmodels` handles this case
+        params_init[d:] = -1*np.ones(q)
+
+    return params_init
+
+
+def _model_complexity(order):
+    (p, d, q) = order
+    # complexity is number of parameters: mu + ar + ma
+    return d + p + q
+
+
+def _batch_trans(p, d, q, nb, x, handle=None):
+    """Apply the stationarity/invertibility guaranteeing transform to batched-parameter vector x."""
+    nvtx_range_push("jones trans")
+
+    if handle is None:
+        handle = cuml.common.handle.Handle()
+
+    Tx = _batched_transform(p, d, q, nb, x, False, handle)
+
+    nvtx_range_pop()
+    return Tx
+
+
+def _batch_invtrans(p, d, q, nb, x, handle=None):
+    """Apply the *inverse* stationarity/invertibility guaranteeing transform to
+       batched-parameter vector x.
+    """
+    nvtx_range_push("jones inv-trans")
+
+    if handle is None:
+        handle = cuml.common.handle.Handle()
+
+    Tx = _batched_transform(p, d, q, nb, x, True, handle)
+
+    nvtx_range_pop()
+    return Tx
+
+
+def _batched_loglike(num_batches, nobs, order, y, np.ndarray[double] x, trans=False, handle=None):
+    cdef vector[double] vec_loglike
+    cdef vector[double] vec_y_cm
+    cdef vector[double] vec_x
+
+    nvtx_range_push("batched loglikelihood")
+    p, d, q = order
+
+    num_params = (p+d+q)
+
+    vec_loglike.resize(num_batches)
+
+    cdef uintptr_t d_y_ptr
+    cdef uintptr_t d_x_ptr
+    cdef uintptr_t d_vs_ptr
+    cdef np.ndarray[double, ndim=2, mode="fortran"] vs = np.zeros(((nobs-d), num_batches), order="F")
```

Comment:
Same comment as the other numpy array, to avoid conflicts this can be defined with:
```
vs = np.zeros(((nobs-d), num_batches), order="F", dtype=np.float64)
```